Demo crashes when outliers are disabled

Looks like I merged #50 prematurely. 

## Reproduction steps
Change
```clojure
(def outliers-enabled false)
```
to
```clojure
(def outliers-enabled false)
```
in `model.clj`, run `(go)` in the REPL and click anywhere in the display window.

## Expected results
The point appears in the display window.

## Actual results
```
java.lang.AssertionError: Assert failed: ["no such subtrace" 2 {"gaussian" {:value -152/25}}]
(not (= val nil))
	at metaprob.state$subtrace.invokeStatic(state.cljc:65)
	at metaprob.state$subtrace.invoke(state.cljc:58)
	at metaprob.trace$trace_direct_subtrace.invokeStatic(trace.cljc:183)
	at metaprob.trace$trace_direct_subtrace.invoke(trace.cljc:182)
	at metaprob.trace$trace_subtrace.invokeStatic(trace.cljc:201)
	at metaprob.trace$trace_subtrace.invoke(trace.cljc:196)
	at curve_fitting.model.trace$outliers$fn__11585.invoke(trace.clj:41)
	at clojure.core$map$fn__5587.invoke(core.clj:2745)
	at clojure.lang.LazySeq.sval(LazySeq.java:40)
	at clojure.lang.LazySeq.seq(LazySeq.java:49)
	at clojure.lang.RT.seq(RT.java:528)
	at clojure.core$seq__5124.invokeStatic(core.clj:137)
	at clojure.core$seq__5124.invoke(core.clj:137)
	at clojure.core$map$fn__5587.invoke(core.clj:2747)
	at clojure.lang.LazySeq.sval(LazySeq.java:40)
	at clojure.lang.LazySeq.seq(LazySeq.java:49)
	at clojure.lang.RT.seq(RT.java:528)
	at clojure.core$seq__5124.invokeStatic(core.clj:137)
	at clojure.core$every_QMARK_.invokeStatic(core.clj:2671)
	at clojure.core$map$step__5598$fn__5599.invoke(core.clj:2764)
	at clojure.lang.LazySeq.sval(LazySeq.java:40)
	at clojure.lang.LazySeq.seq(LazySeq.java:49)
	at clojure.lang.RT.seq(RT.java:528)
	at clojure.core$seq__5124.invokeStatic(core.clj:137)
	at clojure.core$map$fn__5587.invoke(core.clj:2738)
	at clojure.lang.LazySeq.sval(LazySeq.java:40)
	at clojure.lang.LazySeq.seq(LazySeq.java:49)
	at clojure.lang.RT.seq(RT.java:528)
	at clojure.core$seq__5124.invokeStatic(core.clj:137)
	at clojure.core$map$fn__5587.invoke(core.clj:2738)
	at clojure.lang.LazySeq.sval(LazySeq.java:40)
	at clojure.lang.LazySeq.seq(LazySeq.java:49)
	at clojure.lang.RT.seq(RT.java:528)
	at clojure.core$seq__5124.invokeStatic(core.clj:137)
	at clojure.core$map$fn__5591.invoke(core.clj:2749)
	at clojure.lang.LazySeq.sval(LazySeq.java:40)
	at clojure.lang.LazySeq.seq(LazySeq.java:49)
	at clojure.lang.RT.seq(RT.java:528)
	at clojure.core$seq__5124.invokeStatic(core.clj:137)
	at clojure.core$seq__5124.invoke(core.clj:137)
	at curve_fitting.core$draw_clicked_points_BANG_.invokeStatic(core.clj:71)
	at curve_fitting.core$draw_clicked_points_BANG_.invoke(core.clj:60)
	at curve_fitting.core$draw_BANG_.invokeStatic(core.clj:148)
	at curve_fitting.core$draw_BANG_.invoke(core.clj:127)
	at curve_fitting.sketches$applet$fn__13365.invoke(sketches.clj:66)
	at quil.middlewares.fun_mode$wrap_draw_update$quil_draw__1721.invoke(fun_mode.cljc:16)
	at quil.middlewares.safe_fns$wrap_fn$fn__114.invoke(safe_fns.clj:8)
	at quil.applet$_draw.invokeStatic(applet.clj:212)
	at quil.applet$_draw.invoke(applet.clj:210)
	at quil.Applet.draw(Unknown Source)
	at processing.core.PApplet.handleDraw(PApplet.java:2426)
	at quil.Applet.handleDraw(Unknown Source)
	at processing.awt.PSurfaceAWT$12.callDraw(PSurfaceAWT.java:1557)
	at processing.core.PSurfaceNone$AnimationThread.run(PSurfaceNone.java:316)
```

--- a/README.md
+++ b/README.md
@@ -27,5 +27,6 @@ Once you are running the application you can place points by clicking anywhere i
 
 - `c` will clear the screen.
 - `t` will switch "modes" from `prior` to `resampling`, and vice versa
+- `o` will toggle support for outlier detection
 - Digits followed by `return` will set the maximum number of curves.
 - Any other key will refresh the curves.

--- a/deps.edn
+++ b/deps.edn
@@ -9,4 +9,5 @@
                      :sha     "f866d01a8dd5a9354809439912ee430623c4a5c6"}}
 
  :aliases
- {:dev {:extra-paths ["dev"]}}}
+ {:dev {:extra-paths ["dev" "test"]}
+  :test {:extra-paths ["test"]}}}

--- a/dev/user.clj
+++ b/dev/user.clj
@@ -4,4 +4,12 @@
             [integrant.repl.state :refer [system]]
             [curve-fitting.system :as system]))
 
+(defn state
+  []
+  @(:state system))
+
+(defn pprint-state
+  []
+  (clojure.pprint/pprint (state)))
+
 (repl/set-prep! (constantly system/config))

--- a/src/curve_fitting/core.clj
+++ b/src/curve_fitting/core.clj
@@ -113,26 +113,28 @@
         (quil/text text x y)))))
 
 (defn draw-mode!
-  [mode pixel-width pixel-height]
+  [mode outliers? pixel-width pixel-height]
   (quil/rect-mode :corners)
   (quil/text-align :left :bottom)
   (quil/with-fill 0
     (quil/text-size 14) ; pixels
-    (quil/text (case mode
-                 :prior "prior"
-                 :resampling "approximate posterior")
+    (quil/text (str (case mode
+                      :prior "prior"
+                      :resampling "approximate posterior")
+                    " with outliers "
+                    (if outliers? "enabled" "disabled"))
                text-padding
                (- pixel-height text-padding))))
 
 (defn draw!
   "Draws the given state onto the current sketch."
-  [{:keys [mode points curves max-curves digits]} x-scale y-scale pixel-width pixel-height make-opacity-scale]
+  [{:keys [mode outliers? points curves max-curves digits]} x-scale y-scale pixel-width pixel-height make-opacity-scale]
   (let [inverted-x-scale (scales/invert x-scale)
         inverted-y-scale (scales/invert y-scale)
         x-pixel-max (int (/ pixel-width 2))
         x-pixel-min (* -1 x-pixel-max)]
     (quil/background 255)
-    (draw-mode! mode pixel-width pixel-height)
+    (draw-mode! mode outliers? pixel-width pixel-height)
     (draw-curve-count! curves max-curves digits pixel-width pixel-height)
     (let [make-opacity-scale (case mode
                                :resampling resampling/make-opacity-scale

--- a/src/curve_fitting/db.clj
+++ b/src/curve_fitting/db.clj
@@ -8,7 +8,8 @@
    :points     []
    :curves     []
    :digits     []
-   :max-curves 20})
+   :max-curves 20
+   :outliers?  true})
 
 (defn add-curve
   "Adds a curve to the sketch state."
@@ -81,7 +82,21 @@
     (seq digits) (assoc :max-curves (Integer/parseInt (apply str digits)))))
 
 (defn toggle-mode
+  "Sets the simulation to sample from the approximate posterior if it is sampling
+  from the prior and vice versa."
   [{:keys [mode] :as state}]
   (-> state
       (clear-curves)
       (assoc :mode (get {:prior :resampling, :resampling :prior} mode))))
+
+(defn outliers?
+  "Returns true if outliers are enabled, false otherwise."
+  [db]
+  (:outliers? db))
+
+(defn toggle-outliers
+  "Toggles outliers on/off."
+  [db]
+  (-> db
+      (clear-curves)
+      (update :outliers? not)))

--- a/src/curve_fitting/inference.clj
+++ b/src/curve_fitting/inference.clj
@@ -15,7 +15,7 @@
    [metaprob.examples.gaussian :refer :all]))
 
 (define importance-resampling
-  (gen [model-procedure inputs target-trace N]
+  (gen [model-procedure inputs target-trace intervention-trace N]
     ;; generate N candidate traces, called particles, each
     ;; with a score
     (define particles
@@ -25,7 +25,7 @@
                    (define [_ _ score]
                      (infer :procedure model-procedure
                             :inputs inputs
-                            :intervention-trace nil
+                            :intervention-trace intervention-trace
                             :target-trace target-trace
                             :output-trace candidate-trace))
                    [candidate-trace score])))

--- a/src/curve_fitting/model.clj
+++ b/src/curve_fitting/model.clj
@@ -26,7 +26,7 @@
 
 (define add-noise-to-curve
   (gen [curve]
-    ;; Outliers are disabled by default. Enable with an intervention trace.
+    ;; Outliers are enabled by default. Enable with an intervention trace.
     (define outliers-enabled? (flip 1)) ; Are outliers enabled?
     (gen [x]
       (define outlier-point? (flip 0.1))

--- a/src/curve_fitting/model.clj
+++ b/src/curve_fitting/model.clj
@@ -14,8 +14,6 @@
    [metaprob.compositional :as comp]
    [metaprob.examples.gaussian :refer :all]))
 
-(def outliers-enabled true)
-
 (define generate-curve
   (gen []
     (define degree
@@ -26,16 +24,17 @@
       (reduce + (map (gen [n] (* (nth coeffs n) (expt x n)))
                      (range degree))))))
 
-(define outlier?
-  (gen []
-    (and outliers-enabled
-         (flip 0.1))))
-
 (define add-noise-to-curve
   (gen [curve]
+    ;; Outliers are disabled by default. Enable with an intervention trace.
+    (define outliers-enabled? (flip 1)) ; Are outliers enabled?
     (gen [x]
+      (define outlier-point? (flip 0.1))
       (gaussian (curve x)
-                (if (outlier?) 400 0.1)))))
+                (if (and outliers-enabled?
+                         outlier-point?)
+                  400
+                  0.1)))))
 
 (define curve-model
   (gen [xs]

--- a/src/curve_fitting/model.clj
+++ b/src/curve_fitting/model.clj
@@ -39,23 +39,3 @@
 (define curve-model
   (gen [xs]
     (map (add-noise-to-curve (generate-curve)) xs)))
-
-(define get-coeffs-from-trace
-  (gen [tr]
-    (define degree
-      (trace-get
-       (trace-subtrace
-        tr
-        '(1 1 "generate-curve" 0 "degree" "uniform-sample"))))
-
-    (define coeff-subtraces
-      (trace-subtrace
-       tr
-       '(1 1 "generate-curve" 1 "coeffs" "replicate" "map")))
-
-    (map (gen [k]
-           (trace-get
-            (trace-subtrace
-             coeff-subtraces
-             (list k "f" "gaussian"))))
-         (trace-keys coeff-subtraces))))

--- a/src/curve_fitting/model/trace.clj
+++ b/src/curve_fitting/model/trace.clj
@@ -41,13 +41,9 @@
   [ys]
   (fix-points (metaprob/empty-trace) ys))
 
-#_
-(defn add-outlier-target-trace
-  "Given a list of indexes, add an arm to the provided trace structure
-  to fix the points at those indexes outliers or inliers"
-  [traces mode i]
-  (let [tf (if (= mode :inlier) false true)]
-    (assoc-in traces ["map" i 2 "predicate" "outlier?" "then" "flip"] {:value tf})))
+(defn outliers-trace
+  [outliers?]
+  (metaprob/trace-set (metaprob/empty-trace) outliers-enabled-path outliers?))
 
 ;; Points accessors
 
@@ -125,9 +121,3 @@
          (map-indexed (fn [i coefficient]
                         (* coefficient (Math/pow x i))))
          (reduce +))))
-
-#_(let [[_ trace _] (metaprob.interpreters/infer :procedure curve-fitting.model/curve-model
-                                                 :inputs [[1 2 3]]
-                                                 :target-trace (target-trace [1 2 3]))]
-    (outliers trace)
-    [trace (coefficients trace)])

--- a/src/curve_fitting/model/trace.clj
+++ b/src/curve_fitting/model/trace.clj
@@ -2,21 +2,106 @@
   "Functions for manipulating traces from the model."
   (:require [metaprob.builtin :as metaprob]))
 
-(defn target-trace
-  "Returns a target trace that fixes the choices for `y` values in the model to
-  the provided values."
-  [ys]
-  {"map"
-   (into {} (map-indexed (fn [i y]
-                           {i {"gaussian" {:value y}}})
-                         ys))})
+;; Utility
 
+(defn get-subtrace-in
+  [trace path]
+  (if-not (seq path)
+    trace
+    (let [trace-key (first path)]
+      (if-not (metaprob/trace-has-subtrace? trace (list trace-key))
+        nil
+        (get-subtrace-in (metaprob/trace-subtrace trace trace-key)
+                         (rest path))))))
+
+(def points-subtrace-path '("map"))
+(def point-y-path '(1 "gaussian")) ; path from point subtrace to y choice
+(def point-outlier-path '(0 "outlier-point?" "flip")) ; path from point subtrace to outlier choice
+(def outliers-enabled-path '(1 "add-noise-to-curve" 0 "outliers-enabled?" "flip")) ; path from root to outliers-enabled? choice
+
+;; Cosntructors
+
+(defn point-subtrace
+  [{:keys [y outlier?]}]
+  (cond-> (metaprob/empty-trace)
+    (some? y) (metaprob/trace-set point-y-path y)
+    (some? outlier?) (metaprob/trace-set point-outlier-path outlier?)))
+
+(defn fix-points
+  [trace points]
+  (reduce (fn [trace [i point]]
+            (metaprob/trace-set-subtrace trace
+                                         (concat points-subtrace-path (list i))
+                                         (point-subtrace point)))
+          trace
+          (zipmap (range (count points))
+                  points)))
+
+(defn points-trace
+  [ys]
+  (fix-points (metaprob/empty-trace) ys))
+
+#_
 (defn add-outlier-target-trace
   "Given a list of indexes, add an arm to the provided trace structure
   to fix the points at those indexes outliers or inliers"
   [traces mode i]
   (let [tf (if (= mode :inlier) false true)]
     (assoc-in traces ["map" i 2 "predicate" "outlier?" "then" "flip"] {:value tf})))
+
+;; Points accessors
+
+(defn point-count
+  "Returns the number of points in `trace`."
+  [trace]
+  (if-let [points-subtrace (get-subtrace-in trace points-subtrace-path)]
+    (count (metaprob/trace-keys points-subtrace))
+    0))
+
+(defn point-subtraces
+  "Returns a sequence of the point subtraces in `trace`."
+  [trace]
+  (map #(metaprob/trace-subtrace trace (list "map" %))
+       (range (point-count trace))))
+
+(defn point-y
+  "Returns the generated `y` from the point subtrace `subtrace`."
+  [subtrace]
+  (when (metaprob/trace-has? subtrace point-y-path)
+    (metaprob/trace-get subtrace point-y-path)))
+
+(defn point-outlier?
+  "Returns the outlier choice for the point subtrace `subtrace`."
+  [subtrace]
+  (when (metaprob/trace-has? subtrace point-outlier-path)
+    (metaprob/trace-get subtrace point-outlier-path)))
+
+(defn point
+  "Returns the point subtrace `subtrace` as a map."
+  [subtrace]
+  (let [y (point-y subtrace)
+        outlier? (point-outlier? subtrace)]
+    (cond-> {}
+      (some? y) (assoc :y y)
+      (some? outlier?) (assoc :outlier? outlier?))))
+
+(defn points
+  "Returns the points in `trace` as maps."
+  [trace]
+  (map point (point-subtraces trace)))
+
+(defn outliers
+  "Returns each trace's notion of 'outlierness' of each point"
+  [trace]
+  (map point-outlier? (point-subtraces trace)))
+
+(defn outliers-enabled?
+  "Returns true if outliers are enabled in `trace`."
+  [trace]
+  (metaprob/trace-get
+   (metaprob/trace-subtrace trace outliers-enabled-path)))
+
+;; Polynomial
 
 (defn degree
   "Returns the degree from a trace of the model."
@@ -26,36 +111,12 @@
     trace
     '(1 1 "generate-curve" 0 "degree" "uniform-sample"))))
 
-(defn outliers
-  "Returns each trace's notion of 'outlierness' of each point"
-  [trace]
-  (let [outlier-subtraces (if (metaprob/trace-has-subtrace?
-                               trace
-                               '("map"))
-                            (metaprob/trace-subtrace
-                             trace
-                             '("map"))
-                            [])]
-    (map (fn [i]
-           (metaprob/trace-get
-            (metaprob/trace-subtrace
-             outlier-subtraces
-             (list i 2 "predicate" "outlier?" "then" "flip"))))
-         ;; XXX not sure we'll have `then` in all cases?
-         (range (count (metaprob/trace-keys outlier-subtraces))))))
-
 (defn coefficients
-  "Returns the coefficients from a trace of the model."
+  "Returns the coefficients from `trace` in order."
   [trace]
-  (let [coeff-subtraces (metaprob/trace-subtrace
-                         trace
-                         '(1 1 "generate-curve" 1 "coeffs" "replicate" "map"))]
-    (map (fn [i]
-           (metaprob/trace-get
-            (metaprob/trace-subtrace
-             coeff-subtraces
-             (list i "f" "gaussian"))))
-         (range (count (metaprob/trace-keys coeff-subtraces))))))
+  (let [subtrace (get-subtrace-in trace [1 1 "generate-curve" 1 "coeffs" "replicate" "map"])]
+    (map #(metaprob/trace-get subtrace (list % "f" "gaussian"))
+         (range (degree trace)))))
 
 (defn coefficient-function
   [coefficients]
@@ -64,3 +125,9 @@
          (map-indexed (fn [i coefficient]
                         (* coefficient (Math/pow x i))))
          (reduce +))))
+
+#_(let [[_ trace _] (metaprob.interpreters/infer :procedure curve-fitting.model/curve-model
+                                                 :inputs [[1 2 3]]
+                                                 :target-trace (target-trace [1 2 3]))]
+    (outliers trace)
+    [trace (coefficients trace)])

--- a/src/curve_fitting/model/trace.clj
+++ b/src/curve_fitting/model/trace.clj
@@ -5,6 +5,9 @@
 ;; Utility
 
 (defn get-subtrace-in
+  "`clojure.core/get-in`, but for traces. Retrieves the subtrace at the provided
+  key path if it exists. Returns `nil` if any of the keys in the path are
+  missing."
   [trace path]
   (if-not (seq path)
     trace
@@ -22,12 +25,17 @@
 ;; Cosntructors
 
 (defn point-subtrace
-  [{:keys [y outlier?]}]
-  (cond-> (metaprob/empty-trace)
-    (some? y) (metaprob/trace-set point-y-path y)
-    (some? outlier?) (metaprob/trace-set point-outlier-path outlier?)))
+  "Generates a point subtrace that matches the values in `point`. `point` is a
+  map with two optional keys: `:y` and `:outlier?`."
+  [point]
+  (let [{:keys [y outlier?]} point]
+    (cond-> (metaprob/empty-trace)
+      (some? y) (metaprob/trace-set point-y-path y)
+      (some? outlier?) (metaprob/trace-set point-outlier-path outlier?))))
 
 (defn fix-points
+  "Modifies `trace` such that the choices the model makes concerning points match
+  the values in `pionts`. See `point-subtrace` for a description of `points`."
   [trace points]
   (reduce (fn [trace [i point]]
             (metaprob/trace-set-subtrace trace
@@ -38,10 +46,13 @@
                   points)))
 
 (defn points-trace
+  "Returns a trace that fixes the model's outputs to `ys`."
   [ys]
   (fix-points (metaprob/empty-trace) ys))
 
 (defn outliers-trace
+  "Returns a trace that fixes the choice of whether outliers are enabled to
+  `outliers?`."
   [outliers?]
   (metaprob/trace-set (metaprob/empty-trace) outliers-enabled-path outliers?))
 

--- a/src/curve_fitting/sketches.clj
+++ b/src/curve_fitting/sketches.clj
@@ -49,6 +49,9 @@
         (= raw-key \t)
         (db/toggle-mode state)
 
+        (= raw-key \o)
+        (db/toggle-outliers state)
+
         (contains? #{\0 \1 \2 \3 \4 \5 \6 \7 \8 \9} raw-key)
         (db/add-digit state raw-key)
 
@@ -79,20 +82,17 @@
     (try
       (loop []
         (when-not @stop?
-          (let [{old-points :points, old-mode :mode :as old-val} @state
-                curve (case old-mode
-                        :resampling (resampling/sample-curve old-points num-particles)
-                        :prior      (prior/sample-curve old-points))]
-            (swap! state (fn [{new-points :points
-                               new-mode :mode
-                               curves :curves
-                               max-curves :max-curves :as new-val}]
+          (let [{old-points :points, outliers? :outliers?, :as old-state} @state
+                curve (case (:mode old-state)
+                        :resampling (resampling/sample-curve old-points outliers? num-particles)
+                        :prior      (prior/sample-curve old-points outliers?))]
+            (swap! state (fn [{:keys [curves max-curves] :as new-state}]
                            ;; Discard the curve if the points changed while
                            ;; we were working.
-                           (cond-> new-val
-                             (and (= new-points old-points)
-                                  (< (count curves) max-curves)
-                                  (= old-mode new-mode))
+                           (cond-> new-state
+                             (and (= (select-keys old-state [:points :mode :outliers?])
+                                     (select-keys new-state [:points :mode :outliers?]))
+                                  (< (count curves) max-curves))
                              (db/add-curve curve)))))
           (recur)))
       (catch Exception e

--- a/src/curve_fitting/sketches/prior.clj
+++ b/src/curve_fitting/sketches/prior.clj
@@ -5,13 +5,14 @@
             [curve-fitting.scales :as scales]))
 
 (defn sample-curve
-  [points]
+  [points outliers?]
   (let [xs (map :x points)
         ys (map :y points)]
     (let [[_ trace log-score]
           (interpreters/infer :procedure model/curve-model
                               :inputs [xs]
-                              :target-trace (trace/points-trace points))]
+                              :target-trace (trace/points-trace points)
+                              :intervention-trace (trace/outliers-trace outliers?))]
       {:trace trace
        :log-score log-score})))
 

--- a/src/curve_fitting/sketches/prior.clj
+++ b/src/curve_fitting/sketches/prior.clj
@@ -11,7 +11,7 @@
     (let [[_ trace log-score]
           (interpreters/infer :procedure model/curve-model
                               :inputs [xs]
-                              :target-trace (trace/target-trace ys))]
+                              :target-trace (trace/points-trace points))]
       {:trace trace
        :log-score log-score})))
 

--- a/src/curve_fitting/sketches/resampling.clj
+++ b/src/curve_fitting/sketches/resampling.clj
@@ -5,13 +5,16 @@
             [curve-fitting.model.trace :as trace]))
 
 (defn sample-curve
-  [points num-particles]
+  [points outliers? num-particles]
+  (def outliers? outliers?)
   (let [xs (map :x points)
         [trace score] (inference/importance-resampling
                        model/curve-model
                        [xs]
                        (trace/points-trace points)
+                       (trace/outliers-trace outliers?)
                        num-particles)]
+    (def trace trace)
     {:trace trace
      :score (Math/exp score)}))
 

--- a/src/curve_fitting/sketches/resampling.clj
+++ b/src/curve_fitting/sketches/resampling.clj
@@ -7,23 +7,10 @@
 (defn sample-curve
   [points num-particles]
   (let [xs (map :x points)
-        ys (map :y points)
-        y-traces    (trace/target-trace ys)
-        traces      (reduce
-                     (fn [traces [ix mode]]
-                       (case mode
-                         :inlier (trace/add-outlier-target-trace
-                                  traces :inlier ix)
-                         :outlier (trace/add-outlier-target-trace
-                                   traces :outlier ix)
-                         traces))
-                     y-traces
-                     (map-indexed (fn [i p] [i (:outlier-mode p)])
-                                  points))
         [trace score] (inference/importance-resampling
                        model/curve-model
                        [xs]
-                       traces
+                       (trace/points-trace points)
                        num-particles)]
     {:trace trace
      :score (Math/exp score)}))

--- a/test/curve_fitting/model/trace_test.clj
+++ b/test/curve_fitting/model/trace_test.clj
@@ -1,0 +1,45 @@
+(ns curve-fitting.model.trace-test
+  (:require [clojure.test :as test :refer [deftest is testing]]
+            [metaprob.interpreters :refer [infer]]
+            [metaprob.builtin :as metaprob]
+            [curve-fitting.model :as model]
+            [curve-fitting.model.trace :as trace]))
+
+(deftest test-point-subtrace
+  (is (metaprob/trace? (trace/point-subtrace {})))
+  (is (= 1234 (trace/point-y (trace/point-subtrace {:y 1234}))))
+  (is (true? (trace/point-outlier? (trace/point-subtrace {:outlier? true}))))
+  (is (= 1234 (trace/point-y (trace/point-subtrace {:y 1234, :outlier? true}))))
+  (is (true? (trace/point-outlier? (trace/point-subtrace {:y 1234, :outlier? true})))))
+
+(deftest test-point
+  (testing "round trip"
+    (doseq [point [{}
+                   {:y 1}
+                   {:outlier? false}
+                   {:y 2, :outlier? true}]]
+      (is (= point (trace/point (trace/point-subtrace point)))))))
+
+(deftest test-points-trace
+  (is (metaprob/trace? (trace/points-trace [])))
+  (is (metaprob/trace? (trace/points-trace [{}])))
+  (is (metaprob/trace? (trace/points-trace [{:y 1234, :outlier? true}])))
+  (let [trace (trace/points-trace [{:y 1234, :outlier? true}])]
+    (is (= 1234 (trace/point-y (first (trace/point-subtraces trace)))))
+    (is (true? (trace/point-outlier? (first (trace/point-subtraces trace))))))
+  (testing "round trip"
+    (let [point-a {:y 1, :outlier? true}
+          point-b {:y 2, :outlier? false}
+          point-c {:y 3}]
+      (= [point-a] (trace/points (trace/points-trace [point-a])))
+      (= [point-a point-b point-c] (trace/points
+                                    (trace/points-trace [point-a point-b point-c])))))
+  (testing "infer"
+    (let [points [{:x 1, :y 4, :outlier? false}
+                  {:x 2, :y 5, :outlier? true}
+                  {:x 3, :y 6, :outlier? false}]
+          [_ trace _] (infer :procedure model/curve-model
+                             :inputs [(map :x points)]
+                             :target-trace (trace/points-trace points))]
+      (is (= (map #(select-keys % [:y :outlier?]) points)
+             (trace/points trace))))))

--- a/test/curve_fitting/model/trace_test.clj
+++ b/test/curve_fitting/model/trace_test.clj
@@ -43,3 +43,15 @@
                              :target-trace (trace/points-trace points))]
       (is (= (map #(select-keys % [:y :outlier?]) points)
              (trace/points trace))))))
+
+(deftest test-outliers-trace
+  (testing "round trip"
+    (is (true? (trace/outliers-enabled? (trace/outliers-trace true))))
+    (is (false? (trace/outliers-enabled? (trace/outliers-trace false)))))
+  (testing "infer"
+    (let [outliers? true
+          [_ trace _] (infer :procedure model/curve-model
+                             :inputs [[1 2 3]]
+                             :target-trace (metaprob/empty-trace)
+                             :intervention-trace (trace/outliers-trace outliers?))]
+      (is (= outliers? (trace/outliers-enabled? trace))))))

